docs/dynamic-component-loading.md

```diff
@@ -0,0 +1,89 @@
+* `enableUnloading` which will mark the newly create load context to track references and to trigger unload fo the load context and all the assemblies in it when possible.
+
+## High-level description of the implementation
+* Implement a new `AssemblyLoadContext` which will provide the isolation boundary and act as the "root" for the component. It can be enabled for unloading.  
+* The new load context is initialized by specifying the full path to the main assembly of the component to load.  
+* It will look for the `.deps.json` next to that assembly to determine its dependencies. Lack of `.deps.json` will be treated in the same way it is today for executable apps - that is all the assemblies in the same folder will be used as dependencies.
+* Parsing and understanding of the `.deps.json` will be performed by the same host components which do this for executable apps (so same behavior/quirks/bugs, very little code duplication).
+* If the component has `.runtimeconfig.json` it will only be used to verify runtime version and provide probing paths.
+* The load context will determine a list of assemblies similar to TPA and list of native search paths. These will be used to resolve any assembly binding events in the load context.
+* If the load context can't resolve an assembly bind event, it will fallback to the parent load context (the app)
+
+## Important implications and limitations
+* Only framework dependent components will be supported. Self-contained components will not be supported even if there was a way to produce them.
+* The host (the app) can be any configuration (framework dependent or self-contained). The notion of frameworks is completely ignored by this new functionality.
+* All framework dependencies of the component must be resolvable by the app - simply put, the component must use the same frameworks as the app.
+* Components can't add frameworks to the app - the app must "pre-load" all necessary frameworks.
+* By default only framework types will be shared between the app and the component. The component may chose to share more by not including dependencies in the component (`CopyLocal=false`).
+* Pretty much all settings in `.runtimeconfig.json` and `.runtimeconfig.dev.json` will be ignored with the exception of runtime version (probably done through TFM) and additional probing paths.
+
+## TODOs
+* Section about expected tooling experiences - what does it mean to build a "plugin" using .NET SDK and so on.
+* Section about settings/config knobs - env. variables, roll forward settings, command line arguments - which will have effect of component loading and which will be ignored and how.
+* Section about interactions of the new ALC with existing extension points of the binder (various events on `Assembly`, `AppDomain` and `AssemblyLoadContext`, interaction with other ALCs, native asset resolution extension points ...)
+* More details on native asset resolution
+* More details on satellite assembly resolution
```

Comment:
Can we think about plug-in versioning and app compatibility.  Is there a reasonable need to provide a standardized mechanism for plugin to express app requirements.  This might be as simple as adding a deps.json section with convenience API to read it.  

If I was going to create plug-in picker, I would only want to show compatible plug-ins.

Reply:
There are things arleady in place which might help with this. For example:
* The plugin is likely to have a dependency on a an assembly from the app (the one which defines the interface they're going to talk through).
* This dependency is in the `.deps.json` and it has a version - the version which the plugin was compiled with.
* There's existing managed API to read `.deps.json` including this information.

It's not exactly first class support, but at least something is there.

I'll add it to the list of possible improvements to add something more specific.

Reply:
> The plugin is likely to have a dependency on a an assembly from the app

Since it will be provided by the app it will not typically be needed in the plugins deps.json.  So it might want to be declared specially.  Something like 
```json
"pluginAppDependencies" : [ "MyAppPluginModel-3.3.x" ]
```

Reply:
Currently even if a dependency is set to not be included in the output, it will still be listed in the .deps.json. Depending on a situation it might be just a declaration like
```
"Dependency/2.0.1" {},
```
Or it may actually list the assets as well (would contain the actuall .dll path)
But I agree that this doesn't feel like first class support.